### ISSUE TITLE
Login "with email"

### ISSUE THREAD

**Issue by sadaszewski** [NodeBB/NodeBB#10369]
Discourse has one really really good option for helping new websites to gain traction - a password-less, email-based login - https://meta.discourse.org/t/how-do-i-get-the-with-email-login-option/91327 . This is really sweet. Is it possible to add this functionality to NodeBB via a plugin ? If not, any suggestions how to approach this in the main code base ? Thank you in advance.

**Comment by sadaszewski** [comment]
Ideally this should allow to login immediately with the account being created completely transparently if necessary. This would create a kind of "everybody is already registered here" situation which is perfect.

**Comment by julianlam** [comment]
Ah, I was confused by what you were referring to, but what I think you mean is logging in via a "magic url"?

I think this kind of functionality would be best suited as a plugin, and not in core.

**Comment by julianlam** [comment]
It's a neat idea, and while I _personally_ find it cumbersome, I am sure there is an interest in something like this :smile:

**Comment by hello-smile6** [comment]
This would be perfect for creating accounts for users who don't want to register!

**Comment by sadaszewski** [comment]
Magic url is definitely a good term. "Login and *automatic registration* via a magic URL". I highlight the registration part because that one is not necessarily the case in Discourse but I think that a 1-click solution is super important here to attract new users. They are SO lazy and addicted to social networks these days...

**Comment by hello-smile6** [comment]
I actually need auto sign-in with a preset username, so they just click a link and choose a password, but their account exists prior to login. Could that be implemented?

**Comment by sadaszewski** [comment]
@hello-smile6

- password-lessness, 
- automatic username generation/account creation 
- and tight coupling with emails (e.g. deriving the username from the email address)

are THE defining characteristics of *this* plugin. Your idea sounds quite different. As if you had pre-generated accounts and were looking to generate corresponding links to allow users to *claim* the accounts by setting a password. That should be in a different plugin IMHO.

**Comment by hello-smile6** [comment]
> @hello-smile6
> 
> * password-lessness,
> * automatic username generation/account creation
> * tight coupling with emails (e.g. deriving the username from the email address)
> * URL delivery by email
> * Change the "Log in to reply" to simply "Reply" as any mention of a log-in discourages visitors from interacting
> 
> are in my mind THE defining characteristics of _this_ plugin. Your idea sounds quite different. As if you had pre-generated accounts and were looking to generate corresponding links to allow users to _claim_ the accounts by setting a password. That should be in a different plugin IMHO.

[My forum](https://nodebb.9pfs.repl.co/) doesn't have email support, so that wouldn't work.

**Comment by sadaszewski** [comment]
Then perhaps it would be best to keep the "email" part from the original
title of this issue in order to disambiguate.

Email can be a pain to set up on your own but it is doable. Especially if
it is meant to be just an outgoing SMTP server. OpenSMTPd is pretty great.
Then again, you could just use a 3rd party service. And once you have email
support it let's you implement some neat solutions.

**Comment by pitaj** [comment]
What delivery medium would you use for the magic URL besides email?

**Comment by sadaszewski** [comment]
Personally I am interested in email only but I guess one could envisage an SMS as well. That would require a paid service however. Furthermore, SMS could work in an interesting way that is:

1) On a PC the user visits a forum and selects the SMS login
2) The user receives a login URL on their smartphone
3) The user visits the URL on the smartphone
4) The URL authenticates the session on the PC

Well, actually this could be true for any magic URL in general, that is - it would authenticate the session it originates from and not the one it is opened in. Does that make sense?

**Comment by julianlam** [comment]
Heh, if you think email support is spotty in forums (which it is), can you imagine how low uptake of SMS would be?

There's no such thing as a magic URL registration or login if you can't get the actual url to the user, so it'll have to be email. The option to log in with the url will simply be disabled if an emailer is not configured.

**Comment by hello-smile6** [comment]
I don't have an emailer, I have to send emails manually via Gmail.

**Comment by yariplus** [comment]
user-invitations uses a magic url for registration. I could probably change it to do login too.

**Comment by hello-smile6** [comment]
> user-invitations uses a magic url for registration. I could probably change it to do login too.

Please do.

**Comment by pitaj** [comment]
@hello-smile6 FYI, you can set up Gmail in the built in SMTP transport.

**Comment by hello-smile6** [comment]
Tried to, it wanted me to enable 2FA. I don't have a phone, so that's not an option.

**Comment by hello-smile6** [comment]
> @hello-smile6 FYI, you can set up Gmail in the built in SMTP transport.

Also, my forum database is literally public (as are any configuration files) and I don't want to share my email credentials.

**Comment by pitaj** [comment]
Why is your database and config public? Your user sessions can easily be hacked if someone can access the secret in config.json

> Tried to, it wanted me to enable 2FA. I don't have a phone, so that's not an option.

Google wanted you to enable 2FA? If you don't have 2FA enabled you I think you should be able to just put in the email and password. Or you could use a different email service.

**Comment by hello-smile6** [comment]
> Why is your database and config public? Your user sessions can easily be hacked if someone can access the secret in config.json
> 
> > Tried to, it wanted me to enable 2FA. I don't have a phone, so that's not an option.
> 
> Google wanted you to enable 2FA? If you don't have 2FA enabled you I think you should be able to just put in the email and password. Or you could use a different email service.

Yes, people can access the secret, but I don't have the choice. Here's the server, because why not? https://replit.com/@9pfs/NodeBB
If there's a way I can make the database encrypted, make it more reliable, or use environment variables for the config, please let me know.
